poly/src/domain/radix2/fft.rs

```diff
@@ -137,21 +137,53 @@ impl<F: FftField> Radix2EvaluationDomain<F> {
             });
     }
 
+    #[allow(unused)]
     fn io_helper<T: DomainCoeff<F>>(&self, xi: &mut [T], root: F) {
-        let roots = self.roots_of_unity(root);
+        // In the sequential case, we will keep on making the roots cache-aligned,
+        // according to the access pattern that the FFT uses.
+        // It is left as a TODO to implement this for the parallel case
+        let mut roots = self.roots_of_unity(root);
+        let mut root_len = roots.len();
 
         let mut gap = xi.len() / 2;
+        let mut first_iteration = true;
         while gap > 0 {
             // each butterfly cluster uses 2*gap positions
             let chunk_size = 2 * gap;
             let nchunks = xi.len() / chunk_size;
 
-            let butterfly_fn = |(idx, (lo, hi)): (usize, (&mut T, &mut T))| {
+            // Cache align the FFT roots in the sequential case.
+            // In this case, we are aiming to make every root that is accessed one after another,
+            // appear one after another in the list of roots.
+            #[cfg(not(feature = "parallel"))]
+            {
+                // Roots are already cache aligned in the first iteration, so we don't need to do anything.
+                if !first_iteration {
+                    // if the roots are cache aligned in iteration i, then in iteration i+1,
+                    // cache alignment requires selecting every other root.
+                    // (The even powers relative to the current iterations generator)
+                    for i in 1..(root_len / 2) {
+                        roots[i] = roots[i * 2];
+                    }
+                    root_len /= 2;
```

Comment:
Should we also shrink `roots` via `truncate` + `shrink_to_fit`? This would cause re-allocations but can also reduce memory consumption. We don't have to do this at every iteration, maybe every other iteration?

Reply:
I don't see any benefit to doing this. We're in the sequential case, so any re-allocation hurts us here, and the advantage would be that we optimistically freed up some memory for processes running in parallel, of which there would be none